docs/architecture/adr-038-state-listening.md

```diff
@@ -204,60 +204,158 @@ func (rs *Store) CacheMultiStore() types.CacheMultiStore {
-type StreamingService interface {
-    // Stream is the streaming service loop, awaits kv pairs and writes them to a destination stream or file
-    Stream(wg *sync.WaitGroup) error
-    // Listeners returns the streaming service's listeners for the BaseApp to register
-    Listeners() map[types.StoreKey][]store.WriteListener
-    // ABCIListener interface for hooking into the ABCI messages from inside the BaseApp
-    ABCIListener
-    // Closer interface
-    io.Closer
+	// ListenBeginBlock updates the streaming service with the latest BeginBlock messages
+	ListenBeginBlock(blockHeight int64, req []byte, res []byte) error
+	// ListenEndBlock updates the steaming service with the latest EndBlock messages
+	ListenEndBlock(blockHeight int64, req []byte, res []byte) error
+	// ListenDeliverTx updates the steaming service with the latest DeliverTx messages
+	ListenDeliverTx(blockHeight int64, req []byte, res []byte) error
+	// ListenStoreKVPair updates the steaming service with the latest StoreKVPair messages
+	ListenStoreKVPair(blockHeight int64, data []byte) error
+}
+
+// StreamingService struct for registering WriteListeners with the BaseApp and updating the service with the ABCI messages using the hooks
+type StreamingService struct {
+	// Listeners returns the streaming service's listeners for the BaseApp to register
+	Listeners map[types.StoreKey][]store.WriteListener
+	// ABCIListener interface for hooking into the ABCI messages from inside the BaseApp
+	ABCIListener ABCIListener
+	// StopNodeOnErr stops the node when true
+	StopNodeOnErr bool
+}
+```
+
+We will also introduce a new `KVStoreWriter` struct as the writer for use in `StoreKVPairWriteListener`. `KVStoreWriter` will expose `ABCIListener` for streaming state change events via `ABCIListener.ListenStoreKVPair`, a `BlockHeight` func for determining the block height for which the state change events belong to and a boolean flag for determine whether the node needs to be stopped or not.
+
+```go
+// KVStoreWriter is used so that we do not need to update the underlying
+// io.Writer inside the StoreKVPairWriteListener everytime we begin writing
+type KVStoreWriter struct {
+	blockHeight   func() int64
+	listener      ABCIListener
+	stopNodeOnErr bool
+}
+
+// NewKVStoreWriter create an instance of a KVStoreWriter that sends StoreKVPair data to listening service
+func NewKVStoreWriter(
+	listener ABCIListener,
+	stopNodeOnErr bool,
+	blockHeight func() int64,
+) *KVStoreWriter {
+	return &KVStoreWriter{
+		listener:      listener,
+		stopNodeOnErr: stopNodeOnErr,
+		blockHeight:   blockHeight,
+	}
+}
+
+// Write satisfies io.Writer
+func (iw *KVStoreWriter) Write(b []byte) (int, error) {
+	blockHeight := iw.blockHeight()
+	if err := iw.listener.ListenStoreKVPair(blockHeight, b); err != nil {
+		if iw.stopNodeOnErr {
+			panic(err)
+		}
+		return 0, err
+	}
+	return len(b), nil
 }
```

Comment:
@peterbourgon do we need to handle `fire-and-forget` here as well?

I also noticed one more thing (this is pre my involvement with the ADR) and merged into `main`. The error, first propagated to `StoreKVPairWriteListneer.OnWrite`, and then to `Store.OnWrite` is not being handled.

https://github.com/cosmos/cosmos-sdk/blob/fa2cb40892e4653926dd8ff9fb6702e80313af33/store/listenkv/store.go#L151-L155

Reply:
How can you be sure that each Write represents a complete KV store pair?  What constructs one of these types, and where is the implementation of ListenKVPair?

Reply:
If your intent is to send a KV pair to the listener, and you _know_ that a full proto KVPair whatever struct is being serialized and passed to this Write method in a single call — I'm not sure this is true — then you need to parse that proto struct in this method, and refactor ListenStoreKVPair to take those typed values.

Reply:
The implementation...

https://github.com/cosmos/cosmos-sdk/blob/15accd76fc4a2208ad70873277b9f3883a049b58/store/types/listening.go#L32-L47

Reply:
Write is an exported method on an exported type, so it can be called by anything. You can't assume that the OnWrite method of StoreKVPairWriteListener is the only entry point. Probably, KVStoreWriter needs to be removed, and StoreKVPairWriteListener.OnWrite needs to call ListenStoreKVPair directly.

Reply:
Marking it as resolved. The `WriteListener` interface is no longer part of the ADR.